Remind user to add missing phone number

saket opened the pull request:
https://www.pivotaltracker.com/story/show/163448748

Large PR because it includes a database migration and `UpdatePhoneDialog` was duplicated to `AddPhoneDialog`.

saket commented on app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt:
```diff
@@ -450,17 +453,46 @@ class PatientSummaryScreenController @Inject constructor(
+        .map { (number, appointment) -> appointment.updatedAt > number.updatedAt }
+  }
+
+  private fun isMissingPhoneAndShouldBeReminded(patientUuid: UUID): Observable<Boolean> {
```
@vinaysshenoy I'm going back to my older way of calculating when is a patient visiting for the 2nd time:

`When the patient summary screen is opened from search` + `the user has never been reminded before for this patient`.

I don't think it's possible to deduce this otherwise with just the existing data. We will have to persist another value.

saket commented on app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt:
```diff
@@ -450,17 +453,46 @@ class PatientSummaryScreenController @Inject constructor(
-        .take(1)
-        .map { (number) -> { ui: Ui -> ui.showUpdatePhoneDialog(number.patientUuid) } }
+    val showForMissingPhone = screenCreations
+        .filter { it.caller == SEARCH }
```
@vinaysshenoy this controller depends on `PatientSummaryScreenKey#caller == SEARCH` to identify a subsequent visit by a patient. Do you think this will continue to work fine when the summary screen is opened by scanning a BP passport?

vinaysshenoy replied:
Good point. It won't. Should we change it to exclude `NEW_PATIENT` instead?

saket replied:
Yes. Sanchita is going to do it. Please coordinate?

https://github.com/simpledotorg/simple-android/pull/542#discussion_r265457166

saket commented on app/src/androidTest/java/org/simple/clinic/DatabaseMigrationAndroidTest.kt:
```diff
@@ -1348,6 +1348,18 @@ class DatabaseMigrationAndroidTest {
     db_v30.assertColumnCount(tableName, expectedColumnCount)
```
@vinaysshenoy sorry for not breaking this PR into two smaller chunks, but I wanted you to review these two commits together:

- `Remind user to add missing phone number`
- `Add MissingPhoneReminder table to database`

and they constitute the majority of LOC.

saket replied:
They mostly include a duplicated dialog and a DB migration so I'm hoping reviewing it shouldn't be time consuming. Let me know if you still think I should break these two.

vinaysshenoy commented on app/src/main/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialog.kt:
```diff
@@ -0,0 +1,126 @@
+
+  @SuppressLint("CheckResult", "InflateParams")
+  override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+    val layout = LayoutInflater.from(requireContext()).inflate(R.layout.dialog_patientsummary_updatephone, null)
```
The message on the dialog reads "Patient's phone number was previously marked as incorrect" when it should be "Patient's phone number was not entered during registration".

saket replied:
Whoops, fixing this

saket replied:
I had already fixed this, but I think I lost the commit while trying to recover changes from `git reflog`.

vinaysshenoy commented on app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt:
```diff
@@ -450,15 +453,46 @@ class PatientSummaryScreenController @Inject constructor(
-        .filter { (number, appointment) -> appointment.updatedAt > number.updatedAt }
-        .take(1)
-        .map { (number) -> { ui: Ui -> ui.showUpdatePhoneDialog(number.patientUuid) } }
+    val showForMissingPhone = screenCreations
```
Wasn't the trigger for this supposed to be when a new BP is entered? I don't see this description having changed on Zeplin.

saket replied:
Whoops. I got very confused yesterday and forgot to think about this case. Fixing this.

vinaysshenoy commented on app/src/main/java/org/simple/clinic/summary/addphone/MissingPhoneReminder.kt:
```diff
@@ -0,0 +1,28 @@
+data class MissingPhoneReminder(
+    @PrimaryKey
+    val patientUuid: PatientUuid,
+    val reminded: Boolean
```
Even if we do this, we won't need to change the repository API signature since it doesn't actually read this property and controllers will remain untouched.

saket replied:
> I can see us wanting to change from "Never show this dialog once its shown" to something like "Show this dialog once every X days" and this will automatically support both cases.

Sure

saket commented on app/src/test/java/org/simple/clinic/summary/PatientSummaryScreenControllerTest.kt:
```diff
@@ -88,17 +88,17 @@ class PatientSummaryScreenControllerTest {
-        configSubject.firstOrError(),
-        patientSummaryResult,
-        timeFormatter)
+        patientRepository = patientRepository,
```
Agreed

saket commented on app/src/main/java/org/simple/clinic/summary/addphone/MissingPhoneReminder.kt:
```diff
@@ -0,0 +1,28 @@
+    fun save(reminder: MissingPhoneReminder)
+
+    @Query("SELECT * FROM MissingPhoneReminder WHERE patientUuid = :patientUuid")
+    fun get(patientUuid: PatientUuid): Flowable<List<MissingPhoneReminder>>
```
Not yet. They have support for Java8 optional, but until they let us write custom return types, we'll have to continue doing this.

saket commented on app/src/main/java/org/simple/clinic/summary/addphone/MissingPhoneReminderRepository.kt:
```diff
@@ -0,0 +1,21 @@
+import org.simple.clinic.patient.PatientUuid
+import javax.inject.Inject
+
+class MissingPhoneReminderRepository @Inject constructor(private val dao: MissingPhoneReminder.RoomDao) {
```
Oh whoops. Completely forgot this. Fixing this.

saket commented on app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt:
```diff
@@ -450,15 +453,46 @@ class PatientSummaryScreenController @Inject constructor(
-        .take(1)
-        .map { (number) -> { ui: Ui -> ui.showUpdatePhoneDialog(number.patientUuid) } }
+    val showForMissingPhone = screenCreations
+        .filter { it.caller == SEARCH }
```
Agreed. Any suggestion for the name?

vinaysshenoy replied:
`isLatestAppointmentCancelledWithInvalidPhone`? Otherwise, I'd move that filter out into where it's being used so that this filter happens where we are checking if the phone number is incorrect or not.

saket replied:
I'll just move the filter to avoid this long name :p

saket replied:
Actually I can't escape this. I'm just going to use your name suggestion.

saket commented on app/src/main/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialog.kt:
```diff
@@ -0,0 +1,126 @@
+      fragmentManager
+          .beginTransaction()
+          .add(fragment, FRAGMENT_TAG)
+          .commitNowAllowingStateLoss()
```
No, I don't want to care about fragments. If we change this to `commit()`, it'll lead us to a situation where we start seeing `IllegalStateExceptions` because there was a race condition between showing the dialog and stopping the activity.

As an alternative, we can mark the dialog as shown only once its canceled. 

That said, I don't see a strong reason for doing it. The Activity can only stop if the app gets killed midway. What other situations can there be?

vinaysshenoy replied:
Yeah, makes sense. I was probably overthinking this anyway.

saket commented on app/src/main/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialog.kt:
```diff
@@ -0,0 +1,126 @@
+
+  @SuppressLint("CheckResult", "InflateParams")
+  override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+    val layout = LayoutInflater.from(requireContext()).inflate(R.layout.dialog_patientsummary_updatephone, null)
```
I was thinking that we can duplicate the layouts when we need to make changes. Right now they're exactly the same.

vinaysshenoy replied:
Can you see a later comment of mine? The message is different.

saket replied:
I ended up duplicating the layout

saket commented on app/src/main/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialog.kt:
```diff
@@ -0,0 +1,126 @@
+    val saveButton = (dialog as AlertDialog).getButton(DialogInterface.BUTTON_POSITIVE)
+
+    return Observable.merge(dialogCreates(), dialogDestroys, saveClicks(saveButton))
+        .observeOn(Schedulers.io())
```
done

saket commented on app/src/main/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogController.kt:
```diff
@@ -0,0 +1,67 @@
+    return saveUpdatedPhoneNumber(replayedEvents)
+  }
+
+  private fun saveUpdatedPhoneNumber(events: Observable<UiEvent>): Observable<UiChange> {
```
Copy pasta

saket commented on app/src/test/java/org/simple/clinic/summary/PatientSummaryScreenControllerTest.kt:
```diff
@@ -841,4 +845,62 @@ class PatientSummaryScreenControllerTest {
   }
+
+  @Test
+  fun `when screen is opened from search, patient does not have a phone number and the user has never been reminded, then update phone dialog should be shown`() {
```
Whoops, thanks for catching these

vinaysshenoy commented:
@saket There's a failing test, please check?

vinaysshenoy commented on app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt:
```diff
@@ -464,7 +464,12 @@ class PatientSummaryScreenController @Inject constructor(
         }
 
-    val showForMissingPhone = screenCreations
+    val waitTillABpIsRecorded = events
```
The test has to be updated to take this change into account. Right now, it expects the dialog to be shown when the screen is created.

saket replied:
So weird, the tests passed on my local machine. Maybe I missed something. Fixing this.

vinaysshenoy commented on app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:
```diff
@@ -272,12 +280,12 @@ class PatientSummaryScreen(context: Context, attrs: AttributeSet) : RelativeLayo
   fun showBloodPressureUpdateSheet(bloodPressureMeasurementUuid: UUID) {
     val intent = BloodPressureEntrySheet.intentForUpdateBp(context, bloodPressureMeasurementUuid)
-    activity.startActivity(intent)
+    activity.startActivityForResult(intent, REQCODE_BP_ENTRY)
```
This will also trigger the dialog when an older BP is updated. Is this expected?

saket replied:
Hmmm… good catch. Editing of a BP cannot be considered as a "patient visit". I'll remove this.

saket commented:
@vinaysshenoy sorry for that. I remember running tests yesterday and seeing a success message, but it was probably something else. I've updated the tests. 

Additionally, I've also renamed some tests a bit because they were becoming _very_ long.
